### PR TITLE
Add CONDITION_QUEST_OBJECTIVE_COMPLETE

### LINKED ISSUE #17863: [6.x] Core/Conditions: Future Feature CONDITION_QUEST_OBJECTIVE_COMPLETE
**Description**:
Since WoD, phase changes are caused also by single objective completions.
This condition is missing atm. This is just to have an issue opened.

**Current behaviour**: NYI

**Expected behaviour**: Comparable to `CONDITION_QUEST_COMPLETE`
If Objective is complete, condition is met.

**Branch(es)**:  6.x

**TC hash/commit**:  
9b80805d69477102a1ab29925d3c8914d2273476

**TDB version**:  
6.0.3

**Operating system**:  
All

### PR DESCRIPTION
**Pull request by funjoker**
[//]: # (***************************)
[//]: # (** FILL IN THIS TEMPLATE **)
[//]: # (***************************)

**Changes proposed:**

-  Added new Condition to ConditionMgr (thank you @Traesh)
-  Added IsObjectiveComplete bool (thank you @Traesh)

**Target branch(es):** 3.3.5/master

- [x] master

**Issues addressed:** Closes #17863

**Known issues and TODO list:**

### CODE REVIEW

File: src/server/game/Entities/Player/Player.h [hunk 1]
```diff
@@ -1599,6 +1599,7 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
         bool CanShareQuest(uint32 questId) const;
 
         int32 GetQuestObjectiveData(Quest const* quest, int8 storageIndex) const;
+        bool IsQuestObjectiveComplete(Quest const* quest, QuestObjective objective) const;
```
Comment (joschiwald) [inline src/server/game/Entities/Player/Player.h]: QuestObjective const&

**Comment by Aokromes** [comment]
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Conditions/ConditionMgr.cpp:468:42: fatal error: 
      comparison of integers of different signs: 'const int8'
      (aka 'const signed char') and 'const uint32' (aka 'const unsigned int')
      [-Wsign-compare]
                    if (obj.StorageIndex == ConditionValue2)
                        ~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -16637,6 +16597,54 @@ int32 Player::GetQuestObjectiveData(Quest const* quest, int8 storageIndex) const
     return status.ObjectiveData[storageIndex];
 }
 
+bool Player::IsQuestObjectiveComplete(Quest const* quest, QuestObjective const& objective) const
+{
+    switch (objective.Type)
+    {
+    case QUEST_OBJECTIVE_MONSTER:
```
Comment (jackpoz) [inline src/server/game/Entities/Player/Player.cpp]: please keep the original indentation

Comment (funjoker) [inline src/server/game/Entities/Player/Player.cpp]: oh auto formation

**Comment by funjoker** [comment]
@Aokromes 

To Do:

storageIndex can be -1 (Conditions only support uint at the moment)

**Review by Aokromes — CHANGES_REQUESTED** [review]

**Review by Aokromes — APPROVED** [review]

**Comment by Aokromes** [comment]
Conflicting files
src/server/game/Conditions/ConditionMgr.cpp
src/server/game/Conditions/ConditionMgr.h

**Comment by tkrokli** [comment]
AppVeyor failed because I posted commit https://github.com/TrinityCore/TrinityCore/pull/18235/commits/93d2026a1c0850371ac0cab695e649475d9f3128 before the build fix (https://github.com/TrinityCore/TrinityCore/commit/27bf4010b8d12c27d3d8f6df40bff3d9b1a425b1).

**Comment by Shauren** [comment]
`sObjectMgr->GetQuestObjective(questObjectiveId)` is now a thing

**Comment by funjoker** [comment]
@Shauren thanks

File: src/server/game/Globals/ObjectMgr.h [hunk 3]
```diff
@@ -829,6 +829,14 @@ class TC_GAME_API ObjectMgr
             return itr != _questObjectives.end() ? itr->second.first : nullptr;
         }
 
+        std::pair<QuestObjective const*, uint32> const* GetQuestObjectivePairData(uint32 questObjectiveId) const
```
Comment (Shauren) [inline src/server/game/Globals/ObjectMgr.h]: thats not QuestID inside QuestObjective

Comment (funjoker) [inline src/server/game/Globals/ObjectMgr.h]: _questObjectives[obj.ID] = std::make_pair(&obj, qinfo->GetQuestId());

**Comment by funjoker** [comment]
Removed the pair, because it's unecessary if the questid is already in struct.
Would only cause duplicate of data.

**Pull request merged**